### ISSUE TITLE
The supplied InputStream in unnecessarily written into memory

### ISSUE THREAD

**Issue by JannickWeisshaupt** [itplr-kosit/validator#25]
In de/kosit/validationtool/api/InputFactory:154 the supplied input stream is written into memory as ByteArrayOutputStream. It is possible that xRechnungen are arbitrarily large as they contain embedded files. It would be advantageous if the validation tool would perform its operations directly on the stream without completely writing it into memory. This should be supported by saxon.

**Comment by apenski** [comment]
Indeed this is supported by saxon and is definitely a good idea to reduce memory usage of the validator. Changings this requires some rewrite of Input and InputFactory but this should not break the api.


Nevertheless, I tested a litte bit and have trouble at least with the SchemaValidationAction which is not producing  errors with line and col information. I reused the XdmNode parsed in the first step (DocumentParseAction) so that the document is just read once e.g.

```java
final Validator validator = ObjectFactory.createValidator(scenarioType.getSchema());
validator.setErrorHandler(errorHandler);
validator.validate(new DOMSource(NodeOverNodeInfo.wrap(parseDocument.getUnderlyingNode())));
````

With this, I miss the line and column information. When using the byte-array from Input I get them. I'd think they are useful for debugging, so I wouldn't change this until this problem is solved.  Any ideas? 

@rkottmann what do you think?

**Comment by apenski** [comment]
Since Saxon HE has no support for schema, things are more complicated. To run the schema validation, I need to re-read the source. Nevertheless, this reduces overall footprint and is implemented in current master branch.

When using an InputStream, the re-read requires a serialization prior validation. So using a stream somewhat suboptimal: 
* The stream is read when necessary
* must be serialized prior validation

Using an URI based Input is better, since the re-read is based on a new stream. Nevertheless, unless you are using the [ByteArrayInput ](https://github.com/itplr-kosit/validator/blob/master/src/main/java/de/kosit/validationtool/impl/input/ByteArrayInput.java) by using the corresponding[ read in InputFactory](https://github.com/itplr-kosit/validator/blob/master/src/main/java/de/kosit/validationtool/api/InputFactory.java#L229) the target is not read into memory. It's directly parsed to [XdmNode](https://www.saxonica.com/html/documentation/javadoc/net/sf/saxon/s9api/XdmNode.html).

I also implemented support for using [java.xml.transform.Source](https://docs.oracle.com/javase/8/docs/api/javax/xml/transform/Source.html) as an Input. The actual implemenation regarding streams is in [SourceInput](https://github.com/itplr-kosit/validator/blob/master/src/main/java/de/kosit/validationtool/impl/input/SourceInput.java). In future version we may extend for other Source-Implementations as well.

@JannickWeisshaupt Feel free to test this implementation. Any help on spotting errors with this is welcome.

**Comment by apenski** [comment]
Fixed with Release [1.3.0](https://github.com/itplr-kosit/validator/releases/tag/v1.3.0)
